subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java

```diff
@@ -251,19 +251,21 @@ static <T> Value<T> withSideEffect(T value, SideEffect<? super T> sideEffect) {
 
         boolean isMissing();
 
-        @Nullable
-        default SideEffect<? super T> getSideEffect() {
-            return null;
-        }
-
         <S> Value<S> asType();
 
         Value<T> pushWhenMissing(@Nullable DisplayName displayName);
 
         Value<T> addPathsFrom(Value<?> rightValue);
     }
 
-    class PresentWithSideEffect<T> extends Present<T> {
+    interface ValueWithSideEffect<T> extends Value<T> {
```

Comment:
I am not entirely sure, it is the best solution to introduce these methods via the interface hierarchy given the context.

Currently, all "pattern matching" on `Value` happens via calls to methods, such as `isMissing` instead of doing `instanceof` checks. The same observation is applicable to the `ExecutionTimeValue` class. If we introduce side effects via the hierarchy it means that all the pattern matching needs to work with a mix of `isXxx` checks as well as `instanceof` checks. I would argue that keeping the same approach with `boolean`-returning methods keeps the code more consistent.

For example, you can see here configuration caching serialization relies on the `isXxx` checks:
https://github.com/gradle/gradle/blob/6ed98de79c3080333582583bd5d1b6e25a9983c7/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ProviderCodecs.kt#L93

Reply:
What concerns me with adding a new field to `Present` is the potential memory usage increase. Providers and values are used everywhere, at the same time the side effects (for now) are very localized and their absence should not affect memory consumption of most builds.

Reply:
I wouldn't be worried about this reference in `Present`. I think the code would be much simpler without the optimization. I'd do that and then see if we have any problems.